### ISSUE TITLE
JavadocMethod: false positive with validateThrows if catched inside the method

### ISSUE THREAD

**Issue by jdussouillez** [checkstyle/checkstyle#7473]
https://checkstyle.sourceforge.io/config_javadoc.html#JavadocMethod

```
$ javac TestUtils.java
$ echo $?
0
$ cat checkstyle.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
 "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <property name="severity" value="error"/>

    <module name="TreeWalker">
        <module name="JavadocMethod">
            <property name="scope" value="private"/>
            <property name="allowedAnnotations" value="Override"/>
            <property name="validateThrows" value="true"/>
        </module>
    </module>
</module>

$ cat TestUtils.java
package com.myapp.test;

/**
 * Test class
 */
public class TestUtils {

    /**
     * Constructor
     */
    private TestUtils() {
        // "private" to prevent instantiation
    }

    /**
     * Parse a positive integer
     *
     * @param s String to parse
     * @return A positive integer
     * @throws IllegalArgumentException If the number if zero/negative or if an error occured while parsing the string
     */
    private static int parsePositiveInt(final String s) {
        try {
            int value = Integer.parseInt(s);
            if (value <= 0) {
                throw new NumberFormatException(value + " is negative/zero");
            }
            return value;
        } catch (NumberFormatException ex) {
            throw new IllegalArgumentException("Invalid number", ex);
        }
    }
}

$ RUN_LOCALE="-Duser.language=en -Duser.country=US"
$ java $RUN_LOCALE -jar checkstyle-8.28-all.jar -c checkstyle.xml TestUtils.java
Starting audit...
[ERROR] TestUtils.java:26:27: Expected @throws tag for 'NumberFormatException'. [JavadocMethod]
Audit done.
Checkstyle ends with 1 errors.
```

---------------

The `NumberFormatException` exception thrown at line 26 is catched inside the method, so `parsePositiveInt` does not throws `NumberFormatException`. Checkstyle should not report this as an error, it's a false positive.

--------------

**Comment by rnveach** [comment]
Previous version falsely said `IllegalArgumentException` was not needed to be documented

````
$ cat TestClass.java
package com.myapp.test;

/**
 * Test class
 */
public class TestUtils {

    /**
     * Constructor
     */
    private TestUtils() {
        // "private" to prevent instantiation
    }

    /**
     * Parse a positif integer
     *
     * @param s String to parse
     * @return A positive integer
     * @throws IllegalArgumentException If the number if zero/negative or if an error occured while parsing the string
     */
    private static int parsePositiveInt(final String s) {
        try {
            int value = Integer.parseInt(s);
            if (value <= 0) {
                throw new NumberFormatException(value + " is negative/zero");
            }
            return value;
        } catch (NumberFormatException ex) {
            throw new IllegalArgumentException("Invalid number", ex);
        }
    }
}

$ cat TestConfig.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <property name="severity" value="error"/>

    <module name="TreeWalker">
        <module name="JavadocMethod">
            <property name="scope" value="private"/>
            <property name="allowedAnnotations" value="Override"/>
            <property name="validateThrows" value="true"/>
        </module>
    </module>
</module>

$ java -jar checkstyle-8.27-all.jar -c TestConfig.xml TestClass.java
Starting audit...
[ERROR] TestClass.java:20:8: Unused @throws tag for 'IllegalArgumentException'. [JavadocMethod]
Audit done.
Checkstyle ends with 1 errors.
````

**Comment by romani** [comment]
@ jdussouillez , thanks a lot for good issue.
The quickest way to get a fix is to send us PR.
whole logic is at https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java#L697

**Comment by strkkk** [comment]
It should work only if catch contains same type?
Because if there is something like
```
private static int parsePositiveInt(final String s) {
        try {
            int value = Integer.parseInt(s);
            if (value <= 0) {
                throw new NumberFormatException(value + " is negative/zero");
            }
            return value;
        } catch (Exception ex) { // here we catch all
            throw new IllegalArgumentException("Invalid number", ex);
        }
    }
```
what is expected in javadoc? Checkstyle cannot check inheritance chain to understand is catch block suitable or not

P.S. offtop - as for me, catching exception in same method looks weird. Is there useful cases for it? @jdussouillez

**Comment by jdussouillez** [comment]
@strkkk The Javadoc in your example does not require any `@throws` because your method does not throw any checked exceptions.
IMO I would add the `@throws IllegalArgumentException` even if it's not mandatory, because it's useful for the user to know that this exception can be thrown and why.

---

> Is there useful cases for it?

Well it's not something you see everyday but I sometimes use it for some validation/initialization process :

```java
private void init(final int x, final int y) throws InitializationException {
    try {
        if (x <= 0) {
            throw new NegativeValueException("x is negative"); // NegativeValueException inherits InvalidValueException
        }
        if (y > 10_000) {
            throw new OutOfBoundValueException("y is too high"); // OutOfBoundValueException inherits InvalidValueException
        }
        // TODO: add other checks here
    } catch (InvalidValueException | OtherValidationException ex) {
        throw new InitializationException("Error while initializing module", ex);
    }
}
```

**Comment by jdussouillez** [comment]
I just discovered another unexpected behavior which seems linked to the issue. There are also weird results if the method returns a lambda which throws a RuntimeException :

```
$ javac TestUtils2.java
$ echo $?
0
$ cat checkstyle.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
 "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <property name="severity" value="error"/>

    <module name="TreeWalker">
        <module name="JavadocMethod">
            <property name="scope" value="private"/>
            <property name="allowedAnnotations" value="Override"/>
            <property name="validateThrows" value="true"/>
        </module>
    </module>
</module>

$ cat TestUtils2.java
import java.util.function.Function;

/**
 * Test class
 */
public class TestUtils4 {

    /**
     * Constructor
     */
    private TestUtils4() {
        // "private" to prevent instantiation
    }

    /**
     * Returns a function to truncate strings
     *
     * @param maxLength Max length
     * @return A function to truncate string
     */
    private static Function<String, String> createTruncateFunction(final int maxLength) {
        return s -> {
            if (s == null) {
                throw new IllegalArgumentException("Cannot truncate null");
            }
            return s.length() > maxLength ? s.substring(0, maxLength) : s;
        };
    }
}

$ RUN_LOCALE="-Duser.language=en -Duser.country=US"
$ java $RUN_LOCALE -jar checkstyle-8.28-all.jar -c checkstyle.xml TestUtils2.java
Starting audit...
[ERROR] /home/HIFI-FILTER/dussouille/Desktop/test_checkstyle/TestUtils4.java:25:27: Expected @throws tag for 'IllegalArgumentException'. [JavadocMethod]
Audit done.
Checkstyle ends with 1 errors.
```

----

In this case, no `@throws` are expected in the Javadoc because it's not `getTruncateFunction` which throws the `IllegalArgumentException`, but the function it returns. I know this is a bit far-fetched but I'm just thinking of all the possibilities.

**Comment by romani** [comment]
>It should work only if catch contains same type?

Good point.
And as we do not know any details on exception hierarchy, it would be very hard to skip actually caught exceptions.

What if we will use only exception thrown catch block, and ignore all other.

**Comment by wltan** [comment]
> What if we will use only exception thrown catch block, and ignore all other.

This is not foolproof either. It's possible to use `throws` outside of any try/catch blocks entirely, which will need to be included as well.

IMO, the most intuitive solution is to count all `throws` uses, but if it's in a `try` block, then include it if and only if:
1. It is not being caught by a `catch` block (using exact class name only)
2. There is no `finally` block (if there is `finally`, everything from the respective `try` block can be ignored

Regarding https://github.com/checkstyle/checkstyle/issues/7473#issuecomment-574570609 I suppose we could try and look for anonymous classes or lambdas and ignore those, but I'm not sure how much additional work needs to be done on the AST.

The only limitation I can see is the problem with the inheritance chain as mentioned in https://github.com/checkstyle/checkstyle/issues/7473#issuecomment-574529334, which is a known limitation of Checkstyle. I don't see any other way to reduce false positives without introducing false negatives as well.

**Comment by romani** [comment]
> Regarding #7473 (comment) I suppose we could try and look for anonymous classes or lambdas and ignore those, but I'm not sure how much additional work needs to be done on the AST.

it is ok to ignore all throws in lamdas for now. It is better to have false-negative (missed violation) rather then false-positive (unwanted violation). Please do. Regression testing will show what we can improve.

-----

I am thinking on other your questions .... I will reply later.

**Comment by wltan** [comment]
I am on it.

Since we prefer false negative to false positive, I'll start with a very simple rule: ignore all `throws` when it is inside lambda or `try` block. This should remove all false positives within the scope of this issue.

**Comment by rnveach** [comment]
Fix was merged
